node/src/chain/acceptor.rs

```diff
@@ -696,30 +704,99 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
 
             let vm = self.vm.write().await;
 
-            let (stakes, finality) = self.db.read().await.update(|db| {
-                let (txs, verification_output, stake_events) = vm.accept(
-                    prev_header.state_hash,
-                    blk,
-                    &prev_block_voters[..],
-                )?;
-                for spent_tx in txs.iter() {
-                    events.push(TransactionEvent::Executed(spent_tx).into());
-                }
-                est_elapsed_time = start.elapsed();
+            // Events used for archive
+            #[cfg(feature = "archive")]
+            let mut rolling_finality_events = vec![];
+            let (contract_events, finality) =
+                self.db.read().await.update(|db| {
+                    let (txs, verification_output, contract_events) = vm
+                        .accept(
+                            prev_header.state_hash,
+                            blk,
+                            &prev_block_voters[..],
+                        )?;
 
-                assert_eq!(header.state_hash, verification_output.state_root);
-                assert_eq!(header.event_bloom, verification_output.event_bloom);
+                    for spent_tx in txs.iter() {
+                        events
+                            .push(TransactionEvent::Executed(spent_tx).into());
+                    }
+                    est_elapsed_time = start.elapsed();
 
-                let finality =
-                    self.rolling_finality::<DB>(pni, blk, db, &mut events)?;
+                    assert_eq!(
+                        header.state_hash,
+                        verification_output.state_root
+                    );
+                    assert_eq!(
+                        header.event_bloom,
+                        verification_output.event_bloom
+                    );
 
-                let label = finality.0;
-                // Store block with updated transactions with Error and GasSpent
-                block_size_on_disk =
-                    db.store_block(header, &txs, blk.faults(), label)?;
+                    let finality = self.rolling_finality::<DB>(
+                        pni,
+                        blk,
+                        db,
+                        &mut events,
+                        #[cfg(feature = "archive")]
+                        &mut rolling_finality_events,
```

Comment:
If you only need the new final blocks you can use the output of `rolling_finality` method

Indeed
```rust
    /// Returns
    /// - Current accepted block label
    /// - Previous last finalized state root
    /// - List of the new finalized state root
    fn rolling_finality<D: database::DB>(
    ...
    ) -> Result<(Label, Option<RollingFinalityResult>)> {
```